bolt-sidecar/src/builder/fallback/engine_hinter.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    /// Collect hints from the engine API to complete the sealed block.
+    /// This method will keep fetching hints until the payload is valid.
+    pub async fn fetch_payload_from_hints(
+        &self,
+        mut ctx: EngineHinterContext,
+    ) -> Result<SealedBlock, BuilderError> {
+        // The block body can be the same for all iterations, since it only contains
+        // the transactions and withdrawals from the context.
+        let body = ctx.build_block_body();
+
+        // Loop until we get a valid payload from the engine API. On each iteration,
+        // we build a new block header with the hints from the context and fetch the next hint.
+        let max_iterations = 20;
+        let mut iteration = 0;
+        loop {
+            debug!(%iteration, "Fetching hint from engine API");
+
+            // Build a new block header using the hints from the context
+            let header = ctx.build_block_header_with_hints();
+
+            let sealed_hash = header.hash_slow();
+            let sealed_header = SealedHeader::new(header, sealed_hash);
+            let sealed_block = SealedBlock::new(sealed_header, body.clone());
+            let block_hash = ctx.hints.block_hash.unwrap_or(sealed_block.hash());
+
+            // build the new execution payload from the block header
+            let exec_payload = to_alloy_execution_payload(&sealed_block, block_hash);
+
+            // attempt to fetch the next hint from the engine API payload response
+            let hint = self.next_hint(exec_payload, &ctx).await?;
+
+            if matches!(hint, EngineApiHint::ValidPayload) {
+                return Ok(sealed_block);
+            }
+
+            // Populate the new hint in the context and continue the loop
+            ctx.hints.populate_new(hint);
+
+            iteration += 1;
+            if iteration >= max_iterations {
+                return Err(BuilderError::ExceededMaxHintIterations(max_iterations));
+            }
+        }
+    }
+
+    /// Yield the next hint from the engine API by calling `engine_newPayloadV3`
+    /// and parsing the response to extract the hint.
+    ///
+    /// Returns Ok([EngineApiHint::ValidPayload]) if the payload is valid.
+    async fn next_hint(
+        &self,
+        exec_payload: ExecutionPayloadV3,
+        ctx: &EngineHinterContext,
+    ) -> Result<EngineApiHint, BuilderError> {
+        let payload_status = self
+            .engine_client
+            .new_payload_v3(
+                exec_payload,
+                ctx.blob_versioned_hashes.clone(),
+                ctx.parent_beacon_block_root,
+            )
+            .await?;
+
+        let validation_error = match payload_status.status {
+            PayloadStatusEnum::Valid => return Ok(EngineApiHint::ValidPayload),
+            PayloadStatusEnum::Invalid { validation_error } => validation_error,
+            PayloadStatusEnum::Syncing | PayloadStatusEnum::Accepted => {
```

Comment:
Because we don't want the EL to accept this block. According to the engine specs: https://github.com/ethereum/execution-apis/blob/main/src/engine/paris.md#engine_newpayloadv1 the ACCEPTED code MUST be returned only if the following conditions are met:
* all transactions have non-zero length
* the blockHash of the payload is valid
* the payload doesn't extend the canonical chain
* the payload hasn't been fully validated
* ancestors of a payload are known and comprise a well-formed chain

in our case our block extends the canonical chain so it will be marked as VALID.